diff --git a/CMakeLists.txt b/CMakeLists.txt
index a17a261..652e199 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -106,6 +106,7 @@
 # Increased minimum to 3.5 to workaround deprecated backward compatibility
 # since 3.27.
 cmake_minimum_required(VERSION 3.5 FATAL_ERROR)
+cmake_policy(SET CMP0141 NEW)
 project(PCRE2 C)
 set(CMAKE_C_STANDARD 99)
 set(CMAKE_C_STANDARD_REQUIRED TRUE)
@@ -712,7 +713,9 @@ IF(PCRE2_BUILD_PCRE2_8)
     set(targets ${targets} pcre2-posix-static)
 
     IF(MSVC)
-      SET_TARGET_PROPERTIES(pcre2-8-static PROPERTIES OUTPUT_NAME pcre2-8-static)
+      SET_TARGET_PROPERTIES(pcre2-8-static PROPERTIES
+        OUTPUT_NAME pcre2-8-static
+        MSVC_DEBUG_INFORMATION_FORMAT "$<$<CONFIG:Debug>:Embedded>")
       SET_TARGET_PROPERTIES(pcre2-posix-static PROPERTIES OUTPUT_NAME pcre2-posix-static)
     ELSE(MSVC)
       SET_TARGET_PROPERTIES(pcre2-8-static PROPERTIES OUTPUT_NAME pcre2-8)
